### PR TITLE
WIP: inject fake hasher in kube-proxy tests

### PR DESCRIPTION
**Pull request by thockin**
Your PRs reminded me I had started this but never finished.  Rebased onto yours, not complete, just a demonstration.

Better or no?

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -123,20 +123,28 @@ type serviceInfo struct {
...
+		info.serviceFirewallChainName = serviceFirewallChainName(hasher, info.serviceNameString, protocol)
+		info.serviceLBChainName = serviceLBChainName(hasher, info.serviceNameString, protocol)
+
+		return info
+	}
+}
+
+//FIXME:
+type hasher interface {
+	Hash(input ...string) string
```
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go:147]: Get rid of the interface? `type Hasher func(...string) string` ? It would simplify both the real and test implementations

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -123,20 +123,28 @@ type serviceInfo struct {
...
+		info := &serviceInfo{BaseServiceInfo: baseInfo}
+
+		// Store the following for performance reasons.
+		svcName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+		svcPortName := proxy.ServicePortName{NamespacedName: svcName, Port: port.Name}
+		protocol := strings.ToLower(string(info.Protocol()))
+		info.serviceNameString = svcPortName.String()
+		info.servicePortChainName = servicePortChainName(hasher, info.serviceNameString, protocol)
+		info.serviceFirewallChainName = serviceFirewallChainName(hasher, info.serviceNameString, protocol)
+		info.serviceLBChainName = serviceLBChainName(hasher, info.serviceNameString, protocol)
```
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go:139]: (drive-by fix: move the comment to just before the `info.serviceNameString` line, since it's not `svcName`, `svcPortName`, and `protocol` that are being stored for performance reasons).

File: pkg/proxy/iptables/proxier.go [hunk 3]
```diff
@@ -673,44 +686,49 @@ func (proxier *Proxier) OnNodeDelete(node *v1.Node) {
...
 	return encoded[:16]
 }
 
 // servicePortChainName takes the ServicePortName for a service and
 // returns the associated iptables chain.  This is computed by hashing (sha256)
 // then encoding to base32 and truncating with the prefix "KUBE-SVC-".
-func servicePortChainName(servicePortName string, protocol string) utiliptables.Chain {
-	return utiliptables.Chain("KUBE-SVC-" + portProtoHash(servicePortName, protocol))
+func servicePortChainName(hasher hasher, servicePortName string, protocol string) utiliptables.Chain {
+	return utiliptables.Chain("KUBE-SVC-" + hasher.Hash(servicePortName, protocol))
```
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go:710]: due to the aforementioned caching, each of these functions is only called from a single place... maybe get rid of the extra functions, and have:

```
type iptablesChainFunc func(prefix string, ids ...string) utiliptables.Chain

func defaultIPTablesChainMaker(prefix string, ids ...string) utiliptables.Chain {
	buf := bytes.Buffer{}
	for _, id := range ids {
		buf.WriteString(id)
	}
	hash := sha256.Sum256(buf.Bytes())
	encoded := base32.StdEncoding.EncodeToString(hash[:])
	return utiliptables.Chain(prefix+encoded[:16])
}

func newServiceInfo(makeIPTablesChain iptablesChainFunc) func(port *v1.ServicePort, service *v1.Service, baseInfo *proxy.BaseServiceInfo) proxy.ServicePort {
		...
		info.servicePortChainName = makeIPTablesChain("KUBE-SVC-", info.serviceNameString, protocol)
```

File: pkg/proxy/iptables/proxier_test.go [hunk 4]
```diff
@@ -423,6 +427,19 @@ func NewFakeProxier(ipt utiliptables.Interface) *Proxier {
...
 
+type trivialHasher struct{}
+
+func (trivialHasher) Hash(input ...string) string {
+	out := ""
+	for i, in := range input {
+		if i > 0 {
+			out += "-"
+		}
+		out += in
```
Comment (danwinship) [inline pkg/proxy/iptables/proxier_test.go:438]: `out += strings.ToUpper(in)` ?

File: pkg/proxy/iptables/proxier_test.go [hunk 5]
```diff
@@ -480,17 +497,17 @@ COMMIT
...
--A KUBE-SERVICES -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 10.20.30.41 --dport 80 -j KUBE-SVC-XPGD46QRK7WJZT7O
--A KUBE-SVC-XPGD46QRK7WJZT7O -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 10.20.30.41 --dport 80 ! -s 10.0.0.0/24 -j KUBE-MARK-MASQ
--A KUBE-SVC-XPGD46QRK7WJZT7O -m comment --comment ns1/svc1:p80 -j KUBE-SEP-SXIVWICOYRO3J4NJ
--A KUBE-SEP-SXIVWICOYRO3J4NJ -m comment --comment ns1/svc1:p80 -s 10.180.0.1 -j KUBE-MARK-MASQ
--A KUBE-SEP-SXIVWICOYRO3J4NJ -m comment --comment ns1/svc1:p80 -m tcp -p tcp -j DNAT --to-destination 10.180.0.1:80
+-A KUBE-SERVICES -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 10.20.30.41 --dport 80 -j KUBE-SVC-ns1/svc1:p80-tcp
+-A KUBE-SVC-ns1/svc1:p80-tcp -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 10.20.30.41 --dport 80 ! -s 10.0.0.0/24 -j KUBE-MARK-MASQ
+-A KUBE-SVC-ns1/svc1:p80-tcp -m comment --comment ns1/svc1:p80 -j KUBE-SEP-ns1/svc1:p80-tcp
+-A KUBE-SEP-ns1/svc1:p80-tcp -m comment --comment ns1/svc1:p80 -s 10.180.0.1 -j KUBE-MARK-MASQ
+-A KUBE-SEP-ns1/svc1:p80-tcp -m comment --comment ns1/svc1:p80 -m tcp -p tcp -j DNAT --to-destination 10.180.0.1:80
```
Comment (danwinship) [inline pkg/proxy/iptables/proxier_test.go:510]: for consistency with "real" data, the SEP chain names in the parsing tests should include the endpoint IP too. eg, `KUBE-SEP-ns1/svc1:p80-tcp-10.180.0.1:80`.

(actually, some of the chains in the "real" tests are wrong too... I guess you just did search+replace?)

**Review by danwinship — COMMENTED** [review]
So it's nice, but what is the exact goal, because in most places, the un-hashed chain names end up being redundant with a comment, so it doesn't really make anything more understandable... (though, if we wanted to possibly get rid of more comments...)

File: pkg/proxy/iptables/proxier_test.go (hunk 4 above)
Comment (thockin) [reply pkg/proxy/iptables/proxier_test.go]: Yeah, will cleanup format to make this easier to see

File: pkg/proxy/iptables/proxier_test.go (hunk 5 above)
Comment (thockin) [reply pkg/proxy/iptables/proxier_test.go]: I did just s+r.  :)  A little better now.

**Review by thockin — COMMENTED** [review]
Tidied up a bit, all the tests will still fail, but this is for human-review first.  The ONLY goal here is to make test output easier to comprehend.  If it doesn't achieve that goal, let's abandon it.  

Thoughts?  It's less idyllic than I thought it would be, but is it net better?  No offense will be taken :)

**Comment by danwinship** [comment]
> in most places, the un-hashed chain names end up being redundant with a comment, so it doesn't really make anything more understandable...

So yeah, most of the rules end up being like:

    -A KUBE-SERVICES -m comment --comment "ns1/svc1:p80 cluster IP" \
        -m tcp -p tcp -d 172.30.0.41 --dport 80 \
        -j KUBE-SVC-NS1_SVC1_P80_TCP

which doesn't really help much. There is one nice change though, which is that endpoint jumps now include the IP:

    -A KUBE-SVC-NS4_SVC4_P80_TCP -m comment --comment ns4/svc4:p80 \
        -m statistic --mode random --probability 0.5000000000 \
        -j KUBE-SEP-NS4_SVC4_P80_TCP_10.180.0.4:80
    -A KUBE-SVC-NS4_SVC4_P80_TCP -m comment --comment ns4/svc4:p80 \
        -j KUBE-SEP-NS4_SVC4_P80_TCP_10.180.0.5:80`

(maybe change the `.` and `:` to `_` too?)

That would have been slightly useful when looking at dumps to try to see if it picking the right endpoints according to Local traffic policy, I guess.

OTOH, I feel like that would have been useful more often when looking at real iptables dumps than when looking at `proxier_test.go`, so by that argument maybe we should just add the IP to the comment instead...

**Comment by thockin** [comment]
I'm going to leave this alone for now, after code-freeze we can discuss
